File: src/main/java/me/ztowne13/customcrates/listeners/TabCompleteListener.java
package me.ztowne13.customcrates.listeners;

import me.ztowne13.customcrates.SpecializedCrates;
import me.ztowne13.customcrates.commands.sub.SubCommand;
import me.ztowne13.customcrates.crates.Crate;
import me.ztowne13.customcrates.crates.CrateSettings;
import me.ztowne13.customcrates.utils.Utils;
import org.bukkit.Bukkit;
import org.bukkit.World;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by ztowne13 on 2/19/16.
 */
public class TabCompleteListener implements TabCompleter
{
    SpecializedCrates cc;

    public TabCompleteListener(SpecializedCrates cc)
    {
        this.cc = cc;
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command cmd, String alias, String[] args)
    {
        List<String> list = new ArrayList<String>();
        if (sender instanceof Player)
        {
            String cN = cmd.getName();
            if (cN.equalsIgnoreCase("crates") ||
                    cN.equalsIgnoreCase("specializedcrates") ||
                    cN.equalsIgnoreCase("sc") ||
                    cN.equalsIgnoreCase("scrates") ||
                    cN.equalsIgnoreCase("ccrates"))
            {
                if (args.length == 1)
                {
                    list.add("config");
                    list.add("virtualcrates");
                    list.add("givecrate");
                    list.add("givekey");
                    list.add("edit");
                    list.add("errors");
                    list.add("deletecrate");
                    list.add("listcrates");
                    list.add("listhistory");
                    list.add("delallcratetype");
                    list.add("reload");
                    list.add("info");
                    list.add("!");
                    list.add("luckychest");
                    list.add("forceopen");
                    list.add("toggleparticles");
                    list.add("spawncrate");
                    list = Utils.onlyLeaveEntriesWithPref(list, args[0]);
                }
                else
                {
                    if(SubCommand.getMappedAliases().containsKey(args[0]))
                    {
                        args[0] = SubCommand.getMappedAliases().get(args[0]);
                    }

                    if (args[0].equalsIgnoreCase("givekey") || args[0].equalsIgnoreCase("givecrate") ||
                            args[0].equalsIgnoreCase("delallcratetype"))
                    {
                        if (args.length == 2)
                        {
                            for (Crate crates : Crate.getLoadedCrates().values())
                            {
                                list.add(crates.getName());
                            }
                            list = Utils.onlyLeaveEntriesWithPref(list, args[1]);
                        }
                        if (!args[0].equalsIgnoreCase("delallcratetype"))
                        {
                            if (args.length == 3)
                            {
                                list.add("all");
                                for (Player p : Bukkit.getOnlinePlayers())
                                {
                                    list.add(p.getName());
                                }
                                list = Utils.onlyLeaveEntriesWithPref(list, args[2]);
                            }
                            else if (args.length == 4)
                            {
                                list.add("1");
                                list = Utils.onlyLeaveEntriesWithPref(list, "");
                            }
                            else if (args.length == 5)
                            {
                                list.add("-v");
                            }
                        }
                    }
                    else if (args[0].equalsIgnoreCase("edit"))
                    {
                        for (Crate crates : Crate.getLoadedCrates().values())
                        {
                            list.add(crates.getName());
                        }
                        list = Utils.onlyLeaveEntriesWithPref(list, args[1]);
                    }
                    else if (args[0].equalsIgnoreCase("forceopen"))
                    {
                        if(args.length == 2)
                        {
                            for (Crate crates : Crate.getLoadedCrates().values())
                            {
                                list.add(crates.getName());
                            }
                            list = Utils.onlyLeaveEntriesWithPref(list, args[1]);
                        }
                        if(args.length == 3)
                        {
                            list.add("all");
                            for (Player p : Bukkit.getOnlinePlayers())
                            {
                                list.add(p.getName());
                            }

                            list = Utils.onlyLeaveEntriesWithPref(list, args[2]);
                        }
                    }
                    else if (args[0].equalsIgnoreCase(("listhistory")))
                    {
                        if (args.length == 2)
                        {
                            for (Player p : Bukkit.getOnlinePlayers())
                            {
                                list.add(p.getName());
                            }

                            list = Utils.onlyLeaveEntriesWithPref(list, args[1]);
                        }
                        else if (args.length == 3)
                        {
                            list.add("10");
                            list = Utils.onlyLeaveEntriesWithPref(list, "");
                        }
                    }
                    else if (args[0].equalsIgnoreCase("errors"))
                    {
                        for (Crate crates : Crate.getLoadedCrates().values())
                        {
                            list.add(crates.getName());
                        }
                        list = Utils.onlyLeaveEntriesWithPref(list, args[1]);
                    }
                    else if(args[0].equalsIgnoreCase("spawncrate"))
                    {
                        if(args.length == 2)
                        {
                            for (Crate crates : Crate.getLoadedCrates().values())
                            {
                                list.add(crates.getName());
                            }
                            list = Utils.onlyLeaveEntriesWithPref(list, args[1]);
                        } else if(args.length == 3) {
                            for(World world : Bukkit.getWorlds())
                            {
                                list.add(world.getName());
                            }
                            list = Utils.onlyLeaveEntriesWithPref(list, args[2]);
                        } else if(args.length == 4) {
                            list.add("x");
                        } else if(args.length == 5) {
                            list.add("y");
                        } else if(args.length == 6) {
                            list.add("z");
                        }
                    }
                }

            }
            else if(cN.equalsIgnoreCase("rewards"))
            {
                if(args.length >= 1)
                {
                    for (Crate crates : Crate.getLoadedCrates().values())
                    {
                        Player player = (Player) sender;
                        CrateSettings cs = crates.getSettings();

                        if (player.hasPermission(cs.getPermission()) || cs.getPermission().equalsIgnoreCase("no permission"))
                            if (!crates.isMultiCrate())
                                list.add(crates.getName());
                    }

                    list = Utils.onlyLeaveEntriesWithPref(list, args[0]);
                }
            }
        }



        return list;
    }
}

File: src/main/kotlin/me/ztowne13/customcrates/listeners/TabCompleteListener.kt
package me.ztowne13.customcrates.listeners

import org.bukkit.entity.Player
import org.bukkit.Bukkit
import me.ztowne13.customcrates.SpecialisedCrates
import me.ztowne13.customcrates.commands.sub.SubCommand
import me.ztowne13.customcrates.crates.Crate
import org.bukkit.command.*
import java.util.ArrayList

/**
 * Created by ztowne13 on 2/19/16.
 */
class TabCompleteListener(var cc: SpecialisedCrates) : TabCompleter {
    override fun onTabComplete(sender: CommandSender, cmd: Command, alias: String, args: Array<String>): List<String>? {
        var list: MutableList<String>? = ArrayList()
        if (sender is Player) {
            val player = sender
            val cN = cmd.name
            if (cN.equals("crates", ignoreCase = true) ||
                cN.equals("specializedcrates", ignoreCase = true) ||
                cN.equals("sc", ignoreCase = true) ||
                cN.equals("scrates", ignoreCase = true) ||
                cN.equals("ccrates", ignoreCase = true)
            ) {
                if (args.size == 1) {
                    if (player.hasPermission("me.ztowne13.customcrates.admin") || player.hasPermission("specializedcratest.admin")) {
                        list!!.add("config")
                        list.add("virtualcrates")
                        list.add("givecrate")
                        list.add("givekey")
                        list.add("edit")
                        list.add("errors")
                        list.add("deletecrate")
                        list.add("listcrates")
                        list.add("listhistory")
                        list.add("delallcratetype")
                        list.add("reload")
                        list.add("info")
                        list.add("!")
                        list.add("luckychest")
                        list.add("forceopen")
                        list.add("toggleparticles")
                        list.add("spawncrate")
                        list.add("claim")
                    } else {
                        list!!.add("luckychest")
                        list.add("claim")
                    }
                    list = Utils.onlyLeaveEntriesWithPref(list, args[0])
                } else {
                    if (SubCommand.Companion.getMappedAliases().containsKey(args[0])) {
                        args[0] = SubCommand.Companion.getMappedAliases().get(args[0])!!
                    }
                    if (args[0].equals("givekey", ignoreCase = true) || args[0].equals(
                            "givecrate",
                            ignoreCase = true
                        ) ||
                        args[0].equals("delallcratetype", ignoreCase = true)
                    ) {
                        if (args.size == 2) {
                            for (crates in Crate.Companion.getLoadedCrates().values) {
                                list!!.add(crates.name!!)
                            }
                            list = Utils.onlyLeaveEntriesWithPref(list, args[1])
                        }
                        if (!args[0].equals("delallcratetype", ignoreCase = true)) {
                            if (args.size == 3) {
                                list!!.add("all")
                                for (p in Bukkit.getOnlinePlayers()) {
                                    list.add(p.name)
                                }
                                list = Utils.onlyLeaveEntriesWithPref(list, args[2])
                            } else if (args.size == 4) {
                                list!!.add("1")
                                list = Utils.onlyLeaveEntriesWithPref(list, "")
                            } else if (args.size == 5) {
                                list!!.add("-v")
                            }
                        }
                    } else if (args[0].equals("edit", ignoreCase = true)) {
                        for (crates in Crate.Companion.getLoadedCrates().values) {
                            list!!.add(crates.name!!)
                        }
                        list = Utils.onlyLeaveEntriesWithPref(list, args[1])
                    } else if (args[0].equals("forceopen", ignoreCase = true)) {
                        if (args.size == 2) {
                            for (crates in Crate.Companion.getLoadedCrates().values) {
                                list!!.add(crates.name!!)
                            }
                            list = Utils.onlyLeaveEntriesWithPref(list, args[1])
                        }
                        if (args.size == 3) {
                            list!!.add("all")
                            for (p in Bukkit.getOnlinePlayers()) {
                                list.add(p.name)
                            }
                            list = Utils.onlyLeaveEntriesWithPref(list, args[2])
                        }
                    } else if (args[0].equals("listhistory", ignoreCase = true)) {
                        if (args.size == 2) {
                            for (p in Bukkit.getOnlinePlayers()) {
                                list!!.add(p.name)
                            }
                            list = Utils.onlyLeaveEntriesWithPref(list, args[1])
                        } else if (args.size == 3) {
                            list!!.add("10")
                            list = Utils.onlyLeaveEntriesWithPref(list, "")
                        }
                    } else if (args[0].equals("errors", ignoreCase = true)) {
                        for (crates in Crate.Companion.getLoadedCrates().values) {
                            list!!.add(crates.name!!)
                        }
                        list = Utils.onlyLeaveEntriesWithPref(list, args[1])
                    } else if (args[0].equals("spawncrate", ignoreCase = true)) {
                        if (args.size == 2) {
                            for (crates in Crate.Companion.getLoadedCrates().values) {
                                list!!.add(crates.name!!)
                            }
                            list = Utils.onlyLeaveEntriesWithPref(list, args[1])
                        } else if (args.size == 3) {
                            for (world in Bukkit.getWorlds()) {
                                list!!.add(world.name)
                            }
                            list = Utils.onlyLeaveEntriesWithPref(list, args[2])
                        } else if (args.size == 4) {
                            list!!.add("x")
                        } else if (args.size == 5) {
                            list!!.add("y")
                        } else if (args.size == 6) {
                            list!!.add("z")
                        }
                    }
                }
            } else if (cN.equals("rewards", ignoreCase = true)) {
                if (args.size >= 1) {
                    for (crates in Crate.Companion.getLoadedCrates().values) {
                        val cs = crates.settings
                        if (player.hasPermission(cs!!.permission!!) || cs!!.permission.equals(
                                "no permission",
                                ignoreCase = true
                            )
                        ) if (!crates.isMultiCrate) list!!.add(
                            crates.name!!
                        )
                    }
                    list = Utils.onlyLeaveEntriesWithPref(list, args[0])
                }
            }
        }
        return list
    }
}